src/tls/connection.c

```diff
@@ -357,182 +345,173 @@ buffer_read_from_tls (Buffer           *self,
+
+  for (int i = 0; i < sizeof digest_data; i++)
     {
-      https_instance = https_instance_new (peer_der);
-      if (https_instance)
-        {
-          debug (CONNECTION, "ws_socket_name https: created new ws instance %s for new peer certificate", https_instance->socket_name);
-          https_instance->next = https_instances;
-          https_instances = https_instance;
-        }
+      int s = snprintf (result_data, result_size, "%02x", digest_data[i]);
+      assert (s == 2 && 2 < result_size);
+      result_data += 2;
+      result_size -= 2;
     }
 
-  pthread_mutex_unlock (&https_instances_mutex);
+  assert (result_size == 1);
+  assert (result_data[0] == '\0');
 
-  return https_instance ? https_instance->socket_name : NULL;
+  return true;
 }
 
-
-/**
- * connection_init_ws: Connect to a cockpit-ws instance for a new Connection
- */
 static bool
-connection_init_ws (Connection *self)
+connection_connect_to_dynamic_wsinstance (Connection *self)
 {
-  struct sockaddr_un sockaddr = { .sun_family = AF_UNIX };
-  const char *sockname = connection_ws_socket_name (self);
-  int r;
+  struct sockaddr_un addr;
+  char fingerprint[FINGERPRINT_LENGTH + 1];
+
+  assert (self->tls != NULL);
 
-  if (!sockname)
+  if (!get_peer_certificate_fingerprint (self->tls, fingerprint, sizeof fingerprint))
     return false;
 
-  r = snprintf (sockaddr.sun_path, sizeof sockaddr.sun_path,
-                "%s/%s", parameters.wsinstance_sockdir,
-                sockname);
-  assert (r < sizeof sockaddr.sun_path);
+  sockaddr_printf (&addr, "%s/https@%s.sock", parameters.wsinstance_sockdir, fingerprint);
+  debug (CONNECTION, "Connecting dynamic https wsinstance %s:", addr.sun_path);
 
-  debug (CONNECTION, "connection_init_ws: attempting to connect to socket %s", sockaddr.sun_path);
+  /* fast path: the socket already exists, so we can just connect to it */
+  if (connect (self->ws_fd, (struct sockaddr *) &addr, sizeof addr) == 0)
+    return true;
 
-  /* connect to ws instance */
-  self->ws_fd = socket (AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
-  if (self->ws_fd < 0)
+  if (errno != ENOENT && errno != ECONNREFUSED)
+    warn ("connect(%s) failed on the first attempt", addr.sun_path);
+
+  debug (CONNECTION, "  -> failed (%m).  Requesting activation.");
+  /* otherwise, ask for the instance to be started */
+  if (!request_dynamic_wsinstance (fingerprint))
+    return false;
+
+  /* ... and try one more time. */
+  debug (CONNECTION, "  -> trying again");
+  if (connect (self->ws_fd, (struct sockaddr *) &addr, sizeof addr) != 0)
     {
-      warn ("failed to create cockpit-ws client socket");
+      warn ("connect(%s) failed on the second attempt", addr.sun_path);
       return false;
     }
 
-  if (connect (self->ws_fd, (struct sockaddr *) &sockaddr, sizeof sockaddr) < 0)
+  /* otherwise, we're now connected */
+  debug (CONNECTION, "  -> success!");
+  return true;
+}
+
+static bool
+connection_connect_to_static_wsinstance (Connection *self)
```

Comment:
This is only for http connections, and https-without-client-cert is also treated as a dynamic instance, right?  This had me confused for a second.  Would `connection_connect_to_http_wsinstance` be clearer?

Reply:
I thought about this as well. Then also _dynamic → _https. I have no strong opinion, if it's clearer I'm happy to get it renamed.

Reply:
I prefer dynamic because it's what the function is actually doing.  Also, the https-without-client-cert also used to be handled under static (which is where the names come from) and i feel like this is an argument for how minor changes to our approach would continue to work with the static/dynamic names, but if we had http/https names, we'd need to change them...